horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,304 @@
+                                             char* name, bool average) {
+
+  std::string op_name = GetOpName("allreduce", name);
+  auto allreduce_async_fn = [input, output,
+                             op_name](RunContext rctx,
+                                      Callback cb) mutable {
+    DoAllreduce(input, output, op_name, cb);
+  };
+#if HAVE_CUDA
+  auto allreduce_async_cpu_fn =
+      [input, output, op_name](RunContext rctx,
+                               Callback cb) mutable {
+        DoAllreduceCudaOnCPU(input, output, op_name, cb);
+      };
+#endif
+
+#if HAVE_CUDA && !HOROVOD_GPU_ALLREDUCE
+  // Not in-place
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allreduce_async_cpu_fn, input->ctx(),
+                             {input->var()}, {output->var()},
+                             FnProperty::kNormal, 0, "HorovodAllreduce");
+    // In-place
+  } else {
+    Engine::Get()->PushAsync(allreduce_async_cpu_fn, input->ctx(), {},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+  }
+#else
+  // Not in-place
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allreduce_async_fn, input->ctx(), {input->var()},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+    // In-place
+  } else {
+    Engine::Get()->PushAsync(allreduce_async_fn, input->ctx(), {},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllreduce");
+  }
+#endif
+
+  if (average) {
+    *output /= horovod_size();
+  }
+  return 0;
+}
+
+extern "C" int horovod_mxnet_allgather_async(NDArray* input, NDArray* output,
+                                             char* name) {
+
+  std::string op_name = GetOpName("allgather", name);
+  auto allgather_async_fn = [input, output,
+                             op_name](RunContext rctx,
+                                      Callback cb) mutable {
+    DoAllgather(input, output, op_name, cb);
+  };
+#if HAVE_CUDA
+  auto allgather_async_cpu_fn =
+      [input, output, op_name](RunContext rctx,
+                               Callback cb) mutable {
+        DoAllgatherCudaOnCPU(input, output, op_name, cb);
+      };
+#endif
+
+#if HAVE_CUDA && HOROVOD_GPU_ALLGATHER != 'M'
+  // Not in-place
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allgather_async_cpu_fn, Context::CPU(0),
+                             {input->var()}, {output->var()},
+                             FnProperty::kNormal, 0, "HorovodAllgather");
+    // In-place
+  } else {
+    Engine::Get()->PushAsync(allgather_async_cpu_fn, Context::CPU(0), {},
+                             {output->var()}, FnProperty::kNormal, 0,
+                             "HorovodAllgather");
+  }
+#else
+  if (input->var() != output->var()) {
+    Engine::Get()->PushAsync(allgather_async_fn, Context::CPU(0),
```

Comment:
Wanted to double check - in case of allreduce you use `input->ctx()`, why is it always hardcoded to CPU here?

Reply:
You are right. Updated.